### DISCUSSION TITLE
Creating plots in moveHMM

### DISCUSSION THREAD

**Discussion by ToriAlma** [TheoMichelot/moveHMM#35] [Q&A] [answered]
I'm working with moveHMM for the first time and feel a little out of my depth. I'm working with tracking data on African penguins. The observations were irregularly timed and so I used movement rate as step length. I am trying to create several plots from my 2 state model and can't find what I'm looking for in the literature and I am hoping you could shed some light. I would like to:
- create a boxplot of state allocations
- plot step length (velocity) vs state
- create a plot of the x, y points for state 1 and state 2 respectively.
I'm not sure what other info you need but please help. Thanks!

**Comment by TheoMichelot** [comment, marked as answer]
By "boxplot of state allocations", do you mean a boxplot of step lengths grouped by steps?

I am including code to do what you ask, either using base graphics, or using the ggplot2 package. The key function that you need is `viterbi`, which computes the most likely state sequence for a fitted model. Once you have the estimated states, you can use them to subset the data, group by state, etc. The code assumes that the data set (returned by `prepData`) is named `data`, and the model (returned by `fitHMM`) is named `m`.

I hope this helps, please let me know if you have any other questions.

``` R
# Get states and add to data frame
states <- viterbi(m)
data$state <- factor(states)

##################
## Base R plots ##
##################
# Boxplot by state
boxplot(step ~ state, data = data)

# Step lengths against state
plot(as.numeric(data$state), data$step)

# Plot locations by state
plot.default(subset(data, state == 1)[, c("x", "y")])
plot.default(subset(data, state == 2)[, c("x", "y")])

############
## ggplot ##
############
# Boxplot
ggplot(data, aes(x = state, y = step)) + 
    geom_boxplot()

# Step lengths against state
ggplot(data, aes(x = state, y = step)) +
    geom_point()

# Locations by state
ggplot(subset(data, state == 1), aes(x, y)) +
    geom_point() + 
    coord_equal()

ggplot(subset(data, state == 2), aes(x, y)) +
    geom_point() + 
    coord_equal()
```

**Reply by ToriAlma** [reply]
Thank you! I couldn't make the connection on my own and this has been a huge help :)

**Comment by dibeyp** [comment]
Hi Tori

If you have reasonably dense number of positions per day, you could apply
another model to interpolate them into regular time steps and then pass
these to movehmm.
 For this I would suggest you look at "foiegras"

 https://cran.r-project.org/web/packages/foieGras/index.html

This is a robust state space model package which is pretty fast and
reasonably easy to use.

It's worth a look anyway.

Note the first correcting/estimating position and then applying the hidden
Markov model is not 100% statistically "correct" but it often done and to
be honest, there are worse things you could do.

The most correct way to deal with *precise* spatial locations which are
irregular in time would be a continuous time HMM. It can be done but I
don't know if there is a package out there. Theo will know!
HTH
Cheers Toby

**Reply by TheoMichelot** [reply]
Great points Toby, thanks!

Tori, I recently gave a talk on analysing animal movement data in R, available on Youtube: [movement analysis webinar](https://www.youtube.com/watch?v=WELTpbB5BuU). You could have a look around 53:40, where I discuss regularisation of tracking data with foieGras. The code I used is also available on Github: [movement webinar code](https://github.com/eco4cast/Statistical-Methods-Seminar-Series/tree/main/michelot_movement/code).

Another good starting point would be [the foieGras vignette](https://cran.r-project.org/web/packages/foieGras/vignettes/basics.html).

**Reply by ToriAlma** [reply]
Hi Toby, thank you for this feedback, I'll explore this and show it to my thesis supervisor too - looks really interesting. 
Theo, I'll have a look at those videos asap.
Thank you both!
